CI: add to host entity egress end-to-end test

ianvernon opened the pull request:
*  test/runtime: add test for egress to host
* test/helpers: add HostDockerNetwork constant
* test/helpers: add helper function for flushing global connection tracking table
* test: factor out IPs which represent the host
* test/helpers: add helper function for adding IP addresses to VM loopback device
    
Signed-off by: Ian Vernon <ian@cilium.io> 

Related-to: #3436 (contains same commits, but test is separate so that #3436 can get merged before CI is reviewed).

Commit:
test/helpers: add helper function for adding IP addresses to VM loopback device

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test: factor out IPs which represent the host

Move IPs which were previously only used by the CIDR tests out into constants in cons.go.
These IPs are going to be used in upcoming tests for testing traffic to and from the host.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: add test for egress to host

Signed-off-by: Ian Vernon <ian@cilium.io>

Commit:
test/helpers: add helper function for flushing global connection tracking table

Signed-off-by: Ian Vernon <ian@cilium.io>

Commit:
test/helpers: add HostDockerNetwork constant

Signed-off-by: Ian Vernon <ian@cilium.io>

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -633,27 +633,25 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
 		app3Label := "id.app3"
 
-		log.Infof("IPV4 Address Host: %s", ipv4Host)
+		log.Infof("IPV4 Address Host: %s", helpers.IPv4Host)
```
Why not `log.WithFields`, is not clear and less verbose?

ianvernon replied:
The purpose of the log message itself is just to log a value, so I don't see the point of using a field.

eloycoto replied:
I'm not sure If you get me, I suggest to have the four logs and have all in one in a single log with Fields. Like: 
```
vm.logger.WithFields(log.Fields{
"IPv4": helpers.IPv4Host,
"Ipv4AddOtherHost": ipv4OtherHost,
"IPv4NetOther": ipv4OtherNet,
"IPv6": helpers.IPv6Host}).Info("Starting test with this info")
```

For me is clear.

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -716,15 +712,15 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
-		By(fmt.Sprintf("Pinging host IPv4 (%s) from %s (should work) because it is contained within CIDR %s", ipv4Host, helpers.Httpd2, ipv4OtherHost))
-		res = vm.ContainerExec(helpers.Httpd2, helpers.Ping(ipv4Host))
-		res.ExpectSuccess("Unexpected failure pinging host (%s) from %s: %s", ipv4Host, helpers.Httpd2, res.CombineOutput().String())
+		By(fmt.Sprintf("Pinging host IPv4 (%s) from %s (should work) because it is contained within CIDR %s", helpers.IPv4Host, helpers.Httpd2, ipv4OtherHost))
```
For me this By is a bit redundant, the assert message already explain the context of the test no?

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -994,6 +990,74 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+
+		hostDockerContainer := "hostDockerContainer"
+
+		BeforeEach(func() {
```
`BeforeAll` instead of `BeforeEach`

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -994,6 +990,74 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			res.ExpectSuccess("unable to start Docker container with host networking")
+		})
+
+		AfterEach(func() {
```
`AfterAll`

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -994,6 +990,74 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			_, err := vm.PolicyRenderAndImport(policy)
+			Expect(err).To(BeNil(), "Unable to import policy: %s", err)
+
+			areEndpointsReady := vm.WaitEndpointsReady()
```
PolicyRenderAndImport already wait until the endpoints are ready, you can delete this assert message ;-)

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -994,6 +990,74 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			httpd2, err := vm.ContainerInspectNet(helpers.Httpd2)
+			By(fmt.Sprintf("Accessing /public in %s from %s (shouldn't work)", helpers.App2, helpers.App1))
+			failCurl := vm.ContainerExec(helpers.App1, helpers.CurlWithHTTPCode(fmt.Sprintf("http://%s/public", httpd2[helpers.IPv4])))
+			By(fmt.Sprintf("failCurl: %s", failCurl.CombineOutput()))
```
Use `log.debug` or `log.Info` please, `By` is only to provide context for the test

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -994,6 +990,74 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+
+			httpd2, err := vm.ContainerInspectNet(helpers.Httpd2)
+			By(fmt.Sprintf("Accessing /public in %s from %s (shouldn't work)", helpers.App2, helpers.App1))
+			failCurl := vm.ContainerExec(helpers.App1, helpers.CurlWithHTTPCode(fmt.Sprintf("http://%s/public", httpd2[helpers.IPv4])))
```
`CurlWithHTTPCode` does not have --fail option, so if it retuns `[4,5]\d\d` error will not fail. I would sugesst to use CurlFail

aanm commented on test/helpers/cilium.go:
```diff
@@ -799,3 +799,20 @@ func (s *SSHMeta) RestartCilium() error {
+
+// FlushGlobalConntrackTable flushes the global connection tracking table.
+func (s *SSHMeta) FlushGlobalConntrackTable() *CmdRes {
+	return s.ExecCilium("bpf ct flush global")
```
no `sudo`?

ianvernon replied:
`ExecCilium` is a wrapper around `sudo cilium <cmd>`:

https://github.com/cilium/cilium/blob/fad2e59d6b72b04ee8e2f1aedddafa760bca4f55/test/helpers/cilium.go#L59-L63

eloycoto commented on test/helpers/cilium.go:
```diff
@@ -799,3 +799,20 @@ func (s *SSHMeta) RestartCilium() error {
+// AddIPToLoopbackDevice adds the specified IP (assumed to be in form <ip>/<mask>)
+// to the loopback device on s.
+func (s *SSHMeta) AddIPToLoopbackDevice(ip string) *CmdRes {
+	return s.Exec(fmt.Sprintf("sudo ip addr add dev lo %s", ip))
```
`ExecWithSudo`

eloycoto commented on test/helpers/cilium.go:
```diff
@@ -799,3 +799,20 @@ func (s *SSHMeta) RestartCilium() error {
+// RemoveIPFromLoopbackDevice removes the specified IP (assumed to be in form <ip>/<mask>)
+// from the loopback device on s.
+func (s *SSHMeta) RemoveIPFromLoopbackDevice(ip string) *CmdRes {
+	return s.Exec(fmt.Sprintf("sudo ip addr del dev lo %s", ip))
```
`ExecWithSudo`

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -633,27 +633,25 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
 
-		log.Infof("IPV4 Address Host: %s", ipv4Host)
+		log.Infof("IPV4 Address Host: %s", helpers.IPv4Host)
 		log.Infof("IPV4 Address Other Host: %s", ipv4OtherHost)
```
Instead of using log, please using `logger` to have all the logs with the test name. Thanks

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1044,6 +1039,70 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			// Docker container running with host networking is accessible via
+			// the host's IP address. See https://docs.docker.com/network/host/.
+			By(fmt.Sprintf("Accessing /public using Docker container using host networking from %s (should work)", helpers.App1))
+			successCurl := vm.ContainerExec(helpers.App1, helpers.CurlWithHTTPCode(fmt.Sprintf("http://%s/public", hostIP)))
```
Could you delete here the `fmt.Sprintf`, CurlWithHttpCode does not need that: 

```
successCurl := vm.ContainerExec(helpers.App1, helpers.CurlWithHTTPCode("http://%s/public", hostIP))
```

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1044,6 +1039,70 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+
+			httpd2, err := vm.ContainerInspectNet(helpers.Httpd2)
+			By(fmt.Sprintf("Accessing /public in %s from %s (shouldn't work)", helpers.App2, helpers.App1))
+			failCurl := vm.ContainerExec(helpers.App1, helpers.CurlFail(fmt.Sprintf("http://%s/public", httpd2[helpers.IPv4])))
```
Clean the `fmt.Sprintf`:
```
failCurl := vm.ContainerExec(helpers.App1, helpers.CurlFail("http://%s/public", httpd2[helpers.IPv4]))
```

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1044,6 +1039,70 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			failPing := vm.ContainerExec(helpers.App1, helpers.Ping(helpers.App2))
+			failPing.ExpectFail("not able to ping %s", helpers.App2)
+
+			httpd2, err := vm.ContainerInspectNet(helpers.Httpd2)
```
Here a missing Assert on err no?

Commit:
test/helpers: change `ip addr` commands to use `ExecWithSudo`

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: misc. cleanups for host egress test

Clean up curl formatting statements, assertions, logging within test.

Signed-off by: Ian Vernon <ian@cilium.io>
